src/librustc/traits/select.rs

```diff
@@ -1052,12 +1052,39 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
     }
 
     // Treat negative impls as unimplemented
-    fn filter_negative_impls(&self, candidate: SelectionCandidate<'tcx>)
-                             -> SelectionResult<'tcx, SelectionCandidate<'tcx>> {
+    fn filter_negative_and_default_impls<'o>(&self,
+                                             candidate: SelectionCandidate<'tcx>,
+                                             stack: &TraitObligationStack<'o, 'tcx>)
+                                             -> SelectionResult<'tcx, SelectionCandidate<'tcx>> {
+
         if let ImplCandidate(def_id) = candidate {
             if self.tcx().impl_polarity(def_id) == hir::ImplPolarity::Negative {
                 return Err(Unimplemented)
             }
+
+            // if def_id is a default impl and it doesn't implement all the trait items,
+            // the impl doesn't implement the trait.
+            // An `Unimplemented` error is returned only if the default_impl_check is
+            // applicable to the trait predicate or the cause of the predicate is an
+            // `ObjectCastObligation`
+            if self.tcx().impl_is_default(def_id) &&
+               !self.tcx().default_impl_implement_all_methods(def_id){
```

Comment:
As I wrote in a separate comment, I don't think that default impls should be considered impls -- this means we don't even need to make **candidates** for them, I think.

Reply:
The RFC doesn't mention it. It was just my deduction to let them act as regular impls if they are full implemented. I'll remove it.
Concerning the candiates, do you mean moving this check when creating candiates ([select.rs#L1110](https://github.com/giannicic/rust/blob/d6cf35161444e2bb09b115d2670f485c74895d19/src/librustc/traits/select.rs#L1110)) instead of doing it as a filter at the end of the selection process?

Reply:
Right, so, one option might be to [filter the impls out in this loop](https://github.com/giannicic/rust/blob/d6cf35161444e2bb09b115d2670f485c74895d19/src/librustc/traits/select.rs#L1695), but really I suspect that we just don't want them to appear in the result of `for_each_relevant_impl` at all.

In general, I've been trying to think through the idea of doing a "desugaring" step to start. In other words, I'm not happy with how our trait selection code works directly on traits and impls, and I'd prefer for it to work more like chalk -- first lowering traits/impls to some kind of predicate, and then applying those.

I don't necessarily want to block support for default on making that transition, though, but in general we probably want to do a bit of refactoring to better isolate the "things we use to do trait selection" from the "actual impls" that the user wrote.

Reply:
Ok, I'm working on it